mlflow/utils/credentials.py

```diff
@@ -76,29 +77,66 @@ def login(backend="databricks"):
         )
 
 
-def _check_databricks_auth():
-    # Check if databricks credentials are set.
-    try:
-        from databricks.sdk import WorkspaceClient
-    except ImportError:
-        raise ImportError(
-            "Databricks SDK is not installed. To use `mlflow.login()`, please install "
-            "databricks-sdk by `pip install databricks-sdk`."
-        )
+def _validate_databricks_auth(error_queue):
+    from databricks.sdk import WorkspaceClient
 
     try:
-        w = WorkspaceClient()
+        # If no credential is set up, creating `WorkspaceClient` throws an error.
+        workspace_client = WorkspaceClient()
         # If credentials are invalid, `clusters.list()` will throw an error.
-        w.clusters.list()
-        _logger.info("Succesfully signed in Databricks!")
-        # Connect MLflow to Databricks tracking server.
-        from mlflow import set_tracking_uri
-
-        set_tracking_uri("databricks")
-        return True
+        workspace_client.clusters.list()
     except Exception as e:
-        _logger.error(f"Failed to sign in Databricks: {e}")
-        return False
+        # Child process' error is not captured by the parent process, so we need to put the error
+        # into a queue and retrieve it in the parent process.
+        error_queue.put(e)
+
+
+def _connect_to_databricks():
+    """Validate Databricks auth and connect to Databricks.
+
+    This function checks the Databricks authentication, and connects MLflow to Databricks if auth
+    is valid. This function spawns a child process to validate the Databricks auth and automatically
+    times out after 3 seconds. It allows 3 retrys before failing in case of network lagging.
+    """
+    max_trials = 3
+    for _ in range(max_trials):
+        error_queue = multiprocessing.Queue()
+        # Doing auth validation in a child process so that we can kill it after timeout.
+        process = multiprocessing.Process(target=_validate_databricks_auth, args=(error_queue,))
```

Comment:
Multiprocessing module hangs in jupyter notebooks sporadically. Can we use `subprocess.run()` instead?

Reply:
Nice catch! The problem is `subprocess.run()` cannot take a python function. We can switch to using multithreading with the caveat of potentially hanging background thread. None of the solution is perfect though...

Reply:
I am switching to use `subprocess` + databricks CLI as a workaround, which seems to work okay